cpp/ql/test/experimental/query-tests/Security/CWE/CWE-476/semmle/tests/test.cpp

```diff
@@ -0,0 +1,252 @@
+#define NULL ((void*)0)
+typedef unsigned long size_t;
+namespace std {
+  enum class align_val_t : size_t {};
+}
+
+class exception {};
+
+void cleanFunction();
+
+void* operator new(size_t, float);
+void* operator new[](size_t, float);
+void* operator new(size_t, std::align_val_t, float);
+void* operator new[](size_t, std::align_val_t, float);
+void operator delete(void*, float);
+void operator delete[](void*, float);
+void operator delete(void*, std::align_val_t, float);
+void operator delete[](void*, std::align_val_t, float);
+
+struct myData
+{
+    int		sizeInt;
+    char*	buffer;
+};
+
+struct myGlobalData
+{
+    int		sizeInt;
+    myData**	bufMyData;
+};
+
+void allocData(myGlobalData * dataP) {
+    for (size_t i = 0; i < dataP->sizeInt; i++)
+    {
+        dataP->bufMyData[i] = new myData;
+	dataP->bufMyData[i]->sizeInt = 10;
+	dataP->bufMyData[i]->buffer = new char[dataP->bufMyData[i]->sizeInt];
+    }
+}
+
+
+void throwFunction(int a) {
+	if (a == 5)  throw "my exception!";
+}
+void throwFunction2(int a) {
+	if (a == 5) throw exception();
+}
+void funcWork1b() {
+  int a;
+  myGlobalData *valData = new myGlobalData;
+
+  try { 
+  	cleanFunction();
+	throwFunction(a);
+	valData->sizeInt  = 10;
+        valData->bufMyData =  new myData*[valData->sizeInt];
+	cleanFunction();
+	allocData(valData);
+	cleanFunction();
+	 
+  } 
+  catch (...) 
+  {
+    for (size_t i = 0; i < valData->sizeInt; i++)
+    {
+    	delete[] valData->bufMyData[i]->buffer; // BAD
+        delete valData->bufMyData[i];
+    }
+    delete [] valData->bufMyData;
+    delete valData;
+  }
+}
+
+void funcWork1() {
+  int a;
+  myGlobalData *valData = new myGlobalData;
+  valData->sizeInt  = 10;
+  valData->bufMyData =  new myData*[valData->sizeInt];
+  try { 
+  	cleanFunction();
+	throwFunction(a);
+	cleanFunction();
+	allocData(valData);
+	cleanFunction();
+	 
+  } 
+  catch (...) 
+  {
+    for (size_t i = 0; i < valData->sizeInt; i++)
+    {
+	if (valData->bufMyData[i]) 
+       	    delete[] valData->bufMyData[i]->buffer; // GOOD
```

Comment:
`valData->bufMyData[i]` exists due to line 78, but (I think) may contain uninitialized (non-NULL) data if we didn't reach line 83.  I would say this code is still not safe.

Reply:
I need some time to think about this place.

Reply:
initialization `new` is a pretty vague thing.
```
Creates and initializes objects with dynamic storage duration, that is, objects whose lifetime is not necessarily limited by the scope in which they were created.
``` 
but at the same time 
```
The initializer is not optional if

type or new-type is an array of unknown bound
```
but in practice I was not able to get uninitialized data.  
https://www.onlinegdb.com/7SoYOoODS

if you agree, I will add a loop that resets the data after the memory allocation procedure?

Reply:
good example. thanks.